### PR TITLE
Add proper tests for some types

### PR DESCRIPTION
**Pull request by seriyps**
Also, few bugs were fixed:

- encoding and decoding of pg_date (needs verification over postgresql database)
- integer uuid representation

Mostly some basic types covered.

### CODE REVIEW

**Comment by seriyps** [comment]
```
rebar3 cover -v
===> Verifying dependencies...
===> Compiling pg_types
===> Performing cover analysis...
  |------------------------|------------|
  |                module  |  coverage  |
  |------------------------|------------|
  |               pg_json  |      100%  |
  |              pg_range  |        0%  |
  |               pg_inet  |      100%  |
  |       pg_line_segment  |        0%  |
  |            pg_polygon  |        0%  |
  |               pg_int8  |      100%  |
  |               pg_uuid  |       81%  |
  |             pg_record  |        0%  |
  |              pg_jsonb  |      100%  |
  |             pg_circle  |        0%  |
  |               pg_name  |        0%  |
  |         pg_timestampz  |      100%  |
  |            pg_macaddr  |        0%  |
  |               pg_date  |      100%  |
  |               pg_time  |       76%  |
  |              pg_point  |        0%  |
  |          pg_timestamp  |       87%  |
  |                pg_raw  |        0%  |
  |               pg_path  |        0%  |
  |                pg_oid  |        0%  |
  |         pg_bit_string  |        0%  |
  |               pg_int2  |      100%  |
  |             pg_float4  |        0%  |
  |               pg_math  |        0%  |
  |               pg_enum  |        0%  |
  |           pg_interval  |        0%  |
  |            pg_boolean  |      100%  |
  |           pg_tsvector  |        0%  |
  |              pg_array  |        0%  |
  |               pg_line  |        0%  |
  |               pg_int4  |      100%  |
  |             pg_float8  |       30%  |
  |                pg_tid  |        0%  |
  |              pg_types  |        0%  |
  |             pg_hstore  |       83%  |
  |            pg_numeric  |        0%  |
  |------------------------|------------|
  |                 total  |       22%  |
  |------------------------|------------|

```

**Comment by tsloughter** [comment]
Woo, awesome!

**Comment by seriyps** [comment]
Added some more:

```
  |------------------------|------------|
  |                module  |  coverage  |
  |------------------------|------------|
  |               pg_line  |      100%  |
  |                pg_oid  |        0%  |
  |           pg_tsvector  |      100%  |
  |               pg_math  |        0%  |
  |              pg_point  |      100%  |
  |            pg_boolean  |      100%  |
  |              pg_types  |        0%  |
  |               pg_json  |      100%  |
  |               pg_int8  |      100%  |
  |             pg_circle  |      100%  |
  |          pg_timestamp  |       87%  |
  |              pg_array  |        0%  |
  |                pg_tid  |        0%  |
  |               pg_path  |      100%  |
  |               pg_uuid  |       81%  |
  |               pg_enum  |        0%  |
  |             pg_record  |        0%  |
  |            pg_polygon  |      100%  |
  |               pg_inet  |      100%  |
  |             pg_float4  |        0%  |
  |              pg_jsonb  |      100%  |
  |             pg_float8  |       90%  |
  |               pg_date  |      100%  |
  |               pg_name  |       75%  |
  |             pg_hstore  |       83%  |
  |               pg_time  |       76%  |
  |            pg_macaddr  |      100%  |
  |               pg_int2  |      100%  |
  |                pg_raw  |      100%  |
  |         pg_bit_string  |      100%  |
  |               pg_int4  |      100%  |
  |            pg_numeric  |        0%  |
  |              pg_range  |        0%  |
  |           pg_interval  |        0%  |
  |       pg_line_segment  |      100%  |
  |         pg_timestampz  |      100%  |
  |------------------------|------------|
  |                 total  |       35%  |
  |------------------------|------------|
```

I think I'm done now. Remaining ones depend on persistent_term

File: test/prop_uuid.erl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    ?FORALL(Val, proper_types:binary(16),
+            proper_lib:codec(
+              pg_uuid, #{uuid_format => string},
+              iolist_to_binary(strfmt(Val)),
+              fun canonic_str/1)).
+
+prop_integer_codec() ->
+    ?FORALL(Val, proper_types:integer(
+                   -170141183460469231731687303715884105728,
+                   170141183460469231731687303715884105728),
```
Comment (seriyps) [inline test/prop_uuid.erl]: Should UUID be represented as signed integer?

File: test/prop_int.erl [hunk 2]
```diff
@@ -0,0 +1,14 @@
+-module(prop_int).
+-include_lib("proper/include/proper.hrl").
+
+prop_int2_codec() ->
+    ?FORALL(Val, proper_lib:int16(),
+            proper_lib:codec(pg_int2, [], Val)).
```
[the remark this answers is not in the document]
Comment (tsloughter) [reply test/prop_int.erl]: Good point, agreed.

File: src/pg_date.erl [hunk 3]
```diff
@@ -14,10 +14,29 @@ init(_Opts) ->
     {[<<"date_send">>], []}.
 
 encode(Date, _TypeInfo) ->
-    [<<4:?int32>>, encode_date(Date)].
+    [<<4:?int32>>, <<(encode_date(Date)  - ?POSTGRESQL_GD_EPOCH):?int32>>].
 
 decode(<<Date:?int32>>, _TypeInfo) ->
-    calendar:gregorian_days_to_date(Date + ?POSTGRESQL_GD_EPOCH).
+    decode_date(Date + ?POSTGRESQL_GD_EPOCH).
```
[the remark this answers is not in the document]
Comment (tsloughter) [reply src/pg_date.erl]: What do you mean "verify over the real postgresql server"?

**Comment by tsloughter** [comment]
Regarding numeric, I thought @starbelly already did code for handling numerics as tuples instead of floats... but I see it isn't in `pg_numeric` so maybe I lost it...

I'm still not sure what to do about persistent_term being used by types like array. I'm going to toy with recursive resolving today so that types don't have to be looked up in `pg_array`.

File: test/prop_uuid.erl (hunk 1 above)
Comment (tsloughter) [reply test/prop_uuid.erl]: Yea, though pg_types supports integer, string and binary.

File: src/pg_date.erl (hunk 3 above)
[the remark this answers is not in the document]
Comment (seriyps) [reply src/pg_date.erl]: I mean, we are converting datetime to some integer and send it to PostgreSQL; then we read it back as integer and convert back to datetime using the same algorithm. So, what we send and what we will receive will be the same, but what should be verified is that this integer is interpreted as the same `YYYY-MM-DD_HH:mm:ss` as in Erlang.
Say, we have `{{2019, 11, 9}, {19, 59, 0}}` in erlang. We encode it to some integer, send it to postgres, but what if postgres interprets this integer as some other date? Are we sure ituses `gregorian_days_to_date`?

File: test/prop_uuid.erl (hunk 1 above)
Comment (seriyps) [reply test/prop_uuid.erl]: I mean signed vs unsigned integer

**Comment by starbelly** [comment]
@tsloughter hmmm I did, but then i think we decided not to include it. I can send up a PR which adds a clause for it though.

**Comment by tsloughter** [comment]
@seriyps so I have a fix for the use of persistent_term in array and record. It looks like the only catch is that there is some case that there aren't oids available in the record type and it decodes without them because they can be found in the data being decoded https://github.com/tsloughter/pg_types/blob/master/src/pg_record.erl#L67-L69

I don't know when that is an issue that you wouldn't have the oids in `comp_oids` and have to get them from the data as you decode... any idea?

Maybe it isn't something that actually has to be supported?

**Comment by tsloughter** [comment]
Crap, the record decode without knowing the oids before hand is used for tuples. I was just adding a test for `SELECT (NULL, NULL)` and it uses the `decode_tuple_no_oids` function.

**Comment by tsloughter** [comment]
Only took 5 years, hehe, but I merged part of this through a separate PR https://github.com/tsloughter/pg_types/pull/9

**Comment by seriyps** [comment]
Cool, thanks!

**Comment by starbelly** [comment]
> Regarding numeric, I thought @starbelly already did code for handling numerics as tuples instead of floats... but I see it isn't in `pg_numeric` so maybe I lost it...

You didn't lose it, it's here https://github.com/tsloughter/pg_types/blob/master/src/pg_numeric.erl 

But you can't use tuples 😄 You'd have to add an abstraction layer, and maybe that's what you thought I did, instead at some point pgo was trying to use tuples in this way, but I did the work to fix up numerics originally, which was mostly carried over to this repo, here is the original PR to [pgo](https://github.com/erleans/pgo/pull/11)  

> 
> I'm still not sure what to do about persistent_term being used by types like array. I'm going to toy with recursive resolving today so that types don't have to be looked up in `pg_array`.


I'm not sure what the talk about persistent_term here is, but one does have to be very careful with it and also keep in mind that we don't know what people are doing with persistent_term in other libs that an app nor the app itself.

**Comment by tsloughter** [comment]
> I'm not sure what the talk about persistent_term here is, but one does have to be very careful with it and also keep in mind that we don't know what people are doing with persistent_term in other libs that an app nor the app itself.

Here it is used to store type information for mapping oids to types. 

I wanted to keep this library without having any dependency on the other library using it to pass anything but the binary postgres term and the type info for that term. This works everywhere but `pg_record` where there is apparently a case that `comp_type` is `[]`.

I don't remember the details but I think we could fix this by changing the api to accept a map of types.

It just means `pgo` also has to change to update each connection pool record to have and update when types are added this map.

**Comment by starbelly** [comment]
> > I'm not sure what the talk about persistent_term here is, but one does have to be very careful with it and also keep in mind that we don't know what people are doing with persistent_term in other libs that an app nor the app itself.
> 
> Here it is used to store type information for mapping oids to types.
> 
> I wanted to keep this library without having any dependency on the other library using it to pass anything but the binary postgres term and the type info for that term. This works everywhere but `pg_record` where there is apparently a case that `comp_type` is `[]`.
> 
> I don't remember the details but I think we could fix this by changing the api to accept a map of types.
> 
> It just means `pgo` also has to change to update each connection pool record to have and update when types are added this map.

Ahh, so, this would be stored in persistent_term and really never change except for hot code upgrades and restarts, is that correct?

**Comment by tsloughter** [comment]
@starbelly right.
